File: chapechape_client/lib/core/models/reservation_status.dart
/// P1-06 — valeurs Reservation.status canoniques côté Client.
/// Les catégories UI (upcoming / past) se dérivent de ces valeurs.
/// Ne pas confondre avec [paymentStatus].
class ReservationStatusCanon {
  static const pending = 'pending';
  static const awaitingApproval = 'awaiting_approval';
  static const paymentPending = 'payment_pending';
  static const confirmed = 'confirmed';
  static const inStay = 'in_stay';
  static const expired = 'expired';
  static const cancelled = 'cancelled';
  static const completed = 'completed';
  static const refunded = 'refunded';
  static const unknown = 'unknown';

  static const Set<String> canonical = {
    pending,
    awaitingApproval,
    paymentPending,
    confirmed,
    inStay,
    expired,
    cancelled,
    completed,
    refunded,
  };

  /// Alias de caches / anciennes apps. Jamais réémis tels quels.
  static const Map<String, String> aliases = {
    'pending_payment': paymentPending,
    'waiting_payment': paymentPending,
    'payment_required': paymentPending,
    'in_progress': inStay,
    'checked_in': inStay,
    'ongoing': inStay,
    'checked_out': completed,
    'finished': completed,
    'complete': completed,
    'canceled': cancelled,
    'payment_expired': expired,
    'rejected': cancelled,
  };

  static const Set<String> activeUi = {
    pending,
    awaitingApproval,
    paymentPending,
    confirmed,
    inStay,
  };

  static const Set<String> pastUi = {
    expired,
    cancelled,
    completed,
    refunded,
  };

  static String fromApi(String? value) {
    if (value == null || value.trim().isEmpty) return unknown;
    final raw = value.trim().toLowerCase();
    if (canonical.contains(raw)) return raw;
    return aliases[raw] ?? unknown;
  }

  static bool isCanonical(String status) => canonical.contains(status);

  static bool isPaymentTimerStatus(String status) => status == paymentPending;

  static bool isHostApprovalTimerStatus(String status) => status == awaitingApproval;

  static bool isQrEligible(String status) =>
      status == confirmed || status == inStay;

  /// Purpose QR selon le statut canonique (confirmed → checkin, in_stay → checkout).
  static String? stayQrPurpose(String status) {
    final canon = fromApi(status);
    if (canon == confirmed) return 'checkin';
    if (canon == inStay) return 'checkout';
    return null;
  }

  static bool isUpcoming(String status, DateTime checkIn, DateTime now) {
    if (pastUi.contains(status)) return false;
    return activeUi.contains(status) && !checkIn.isBefore(now);
  }

  static bool isPast(String status, DateTime checkOut, DateTime now) {
    if (pastUi.contains(status)) return true;
    return checkOut.isBefore(now);
  }
}

File: backend/src/constants/reservation-status.js
/**
 * P1-06 — source canonique Reservation.status.
 * Ne pas confondre avec Payment.status / paymentStatus, ExternalReservation.status,
 * AvailabilityBlock.status, Residence.status, User.status.
 */

const RESERVATION_STATUS = Object.freeze({
  PENDING: 'pending',
  AWAITING_APPROVAL: 'awaiting_approval',
  PAYMENT_PENDING: 'payment_pending',
  CONFIRMED: 'confirmed',
  IN_STAY: 'in_stay',
  EXPIRED: 'expired',
  CANCELLED: 'cancelled',
  COMPLETED: 'completed',
  REFUNDED: 'refunded',
});

const RESERVATION_STATUS_VALUES = Object.freeze(Object.values(RESERVATION_STATUS));

/** Occupent réellement l'inventaire. Sous-ensemble métier, pas un second lifecycle. */
const ACTIVE_BLOCKING_STATUSES = Object.freeze([
  RESERVATION_STATUS.PENDING,
  RESERVATION_STATUS.AWAITING_APPROVAL,
  RESERVATION_STATUS.PAYMENT_PENDING,
  RESERVATION_STATUS.CONFIRMED,
  RESERVATION_STATUS.IN_STAY,
]);

const TERMINAL_STATUSES = Object.freeze([
  RESERVATION_STATUS.EXPIRED,
  RESERVATION_STATUS.CANCELLED,
  RESERVATION_STATUS.COMPLETED,
  RESERVATION_STATUS.REFUNDED,
]);

/**
 * Alias d'entrée (anciennes apps / caches). Jamais persistés.
 * `rejected` n'est PAS un alias : le rejet passe par l'endpoint dédié
 * (cancelled + rejectedByHost).
 */
const RESERVATION_STATUS_INPUT_ALIASES = Object.freeze({
  pending_payment: RESERVATION_STATUS.PAYMENT_PENDING,
  waiting_payment: RESERVATION_STATUS.PAYMENT_PENDING,
  payment_required: RESERVATION_STATUS.PAYMENT_PENDING,
  in_progress: RESERVATION_STATUS.IN_STAY,
  checked_in: RESERVATION_STATUS.IN_STAY,
  ongoing: RESERVATION_STATUS.IN_STAY,
  checked_out: RESERVATION_STATUS.COMPLETED,
  finished: RESERVATION_STATUS.COMPLETED,
  complete: RESERVATION_STATUS.COMPLETED,
  canceled: RESERVATION_STATUS.CANCELLED,
  payment_expired: RESERVATION_STATUS.EXPIRED,
});

const LEGACY_ALIAS_VALUES = Object.freeze(Object.keys(RESERVATION_STATUS_INPUT_ALIASES));

function normalizeReservationStatusInput(value) {
  if (value == null) return value;
  const raw = String(value).trim().toLowerCase();
  if (RESERVATION_STATUS_VALUES.includes(raw)) return raw;
  if (Object.prototype.hasOwnProperty.call(RESERVATION_STATUS_INPUT_ALIASES, raw)) {
    return RESERVATION_STATUS_INPUT_ALIASES[raw];
  }
  return raw;
}

function isCanonicalReservationStatus(value) {
  return RESERVATION_STATUS_VALUES.includes(value);
}

function isLegacyReservationStatusAlias(value) {
  if (value == null) return false;
  return Object.prototype.hasOwnProperty.call(
    RESERVATION_STATUS_INPUT_ALIASES,
    String(value).trim().toLowerCase()
  );
}

module.exports = {
  RESERVATION_STATUS,
  RESERVATION_STATUS_VALUES,
  ACTIVE_BLOCKING_STATUSES,
  TERMINAL_STATUSES,
  RESERVATION_STATUS_INPUT_ALIASES,
  LEGACY_ALIAS_VALUES,
  normalizeReservationStatusInput,
  isCanonicalReservationStatus,
  isLegacyReservationStatusAlias,
};
